src/Components/Endpoints/src/Builder/RazorComponentEndpointConventionBuilder.cs

```diff
@@ -15,21 +15,43 @@ public class RazorComponentEndpointConventionBuilder : IEndpointConventionBuilde
 {
     private readonly object _lock;
     private readonly ComponentApplicationBuilder _builder;
+    private readonly List<IComponentRenderMode> _renderModes;
     private readonly List<Action<EndpointBuilder>> _conventions;
     private readonly List<Action<EndpointBuilder>> _finallyConventions;
 
     internal RazorComponentEndpointConventionBuilder(
         object @lock,
         ComponentApplicationBuilder builder,
+        List<IComponentRenderMode> renderModes,
         List<Action<EndpointBuilder>> conventions,
         List<Action<EndpointBuilder>> finallyConventions)
     {
         _lock = @lock;
         _builder = builder;
+        _renderModes = renderModes;
         _conventions = conventions;
         _finallyConventions = finallyConventions;
     }
 
+    /// <summary>
+    /// Gets the <see cref="ComponentApplicationBuilder"/> that is used to build the endpoints.
+    /// </summary>
+    public ComponentApplicationBuilder ApplicationBuilder => _builder;
+
+    /// <summary>
+    /// Restricts the <see cref="IComponentRenderMode"/> available for these endpoints.
+    /// </summary>
+    /// <param name="renderModes">The render modes the components rendering from these endpoints are
+    /// allowed to use.
+    /// </param>
+    /// <returns>The <see cref="RazorComponentEndpointConventionBuilder"/>.</returns>
+    public RazorComponentEndpointConventionBuilder SetRenderModes(params IComponentRenderMode[] renderModes)
```

Comment:
I suggest changing this to `AllowServer` and `AllowWebAssembly`. The current API is awkward because:

 * It implies you can control whether or not "auto" is allowed, which doesn't really make sense
 * It implies you can control whether or not prerendering is allowed, which is not true. Similar for any other parameters added to the modes, such as auto policy.

If these APIs really mean "I want to register the endpoints for Server/WebAssembly" then it would be clearer for the API to be phrased in those terms.

Reply:
This is used to restrict the set of render modes an application can render on. It doesn't make sense to pass auto because "auto" pretty much means "anything".

Webassembly and Server will have a `SetWebAssemblyRenderingMode` and `SetBlazorServerRenderingMode` to constraint the usage to webassembly only.

None of this is necessary if you don't have more than one app or you are not registering a given mode via (AddServerComponents or AddWebassemblyComponents).

An alternative to this, would be to have a different top level API, MapServerRazorComponents<App> and MapWebassemblyRazorComponents<App>.

Reply:
To be clear, this works as follows:
* 1 app:
  * Call AddRazorComponents and optionally, AddWebassemblyComponents, AddServerComponents. 
  * MapRazorComponents automatically detects what endpoints to wire-up based on the discovered component render modes.
  * If you have server components but don't call AddServerComponents, we throw an exception and let you know that it is missing. Similarly, with Webassembly.
* n-apps:
  * Each app gets configured based on the components rendered modes discovered in their dependency graph.
  * You can optionally restrict the render modes available. Components with incompatible render modes won't work as we won't plug in the required endpoints.

Reply:
I might be missing something but it's difficult to follow because the PR currently builds the list of `_renderModes` but then seemingly doesn't do anything with that information. I'm not certain what it's going to affect.

> If you have server components but don't call AddServerComponents, we throw an exception and let you know that it is missing. Similarly, with Webassembly.

What if you're not going to use those specific components? They might come from some package where you want to use one of its components but not others. Or do you mean this exception happens when the component is actually rendered at runtime?

My mental model might be different from yours but I expected the thing the developer cares about is "have I enabled the endpoints for Server (the hub) or WebAssembly (the static files) within a given branch of the middleware pipeline?" I didn't think that would need to be tied to having mapped endpoints for some components of that type.

A converse situation would be if you want to enable running some things on WebAssembly but do *not* want those components to ever run on SSR, so you don't want to map any endpoints for them.

Reply:
I have clarified the scenarios in the description for the PR. The mechanics is that when you have a single `MapRazorComponents<T>` whatever services you configure and whatever your components have declared determine what render modes you get. If you are missing calls to `AddXComponents` you'll get an exception explaining what to do.

If you don't want to enable a specific mode, you can choose between server, webassembly or static rendering modes, and we'll respect that (and avoid wiring up the endpoints). Your components will prerender but won't be interactive.

If you have multiple branches and want some parts to be server, you want a branch to be server, a branch to be webassembly, or a branch to be static, you can set the render mode explicitly.